### ISSUE TITLE
Cross-Vendor Multi-GPU Support via Vulkan Backend

### ISSUE THREAD

**Issue by dproworld** [comfyanonymous/ComfyUI#4170]
I realise that this is a far stretch, but a vulkan backend which would allow for multi-gpu inference would be awesome. With new models like Flux, inference requires more than 12gb at fp8. Vulkan would solve cross nvidia-AMD setups and had been implemented in llama.cpp (would be different, but its still torch). I got llama inference working on my 3060 12gb rx 6750xt setup with vulkan, so it would be great if something like this was possible. 

While similar requests have been brought up in the past, it was unclear to me whether or not they were meant for multi-gpu inference for singular images in the interest of vram.

**Comment by voidastro4** [comment]
LLVM-IREE ( https://github.com/iree-org/iree ) has a vulkan backend functional for pytorch, onnx, jax.
They have a working stable diffusion example: https://github.com/nod-ai/SHARK

**Comment by abdulhakam** [comment]
I recently tried SD1.5 (not SDXL) on vulkan using kobold cpp as a test, it was fast enough, a (little) bit slower than CPU(since its an iGPU (intel 620HD)), but overall it saved time since I was still able to browse the internet without any hanging up or crashes. I would really love it on comfyUI, it might help supporting all kinds of GPU's which support vulkan.

**Comment by dproworld** [comment]
Yeah, vulkan is awesome. LLVM-IREE seems promising, but integrated vulkan backend to comfyui would provide better results because of libs available with comfyui.

**Comment by waltercool** [comment]
Using Vulkan would be great, ROCM works nice, but their KFD breaks too easy, making whole AMDGPU operation unusable unless you reload the module.

**Comment by morphles** [comment]
I'll second that Vulkan would be be amazing even if potentially less performant. Vulkan seems to be  simpler to setup than juggling ROCM or cuda versions, and should work on many more vendors, potentially even on mobile vie termux or such. If LLMs can work on it SD should also be able to work :)

**Comment by Iemand005** [comment]
Vulkan would be a really good alternative for a backend that just works on AMD, Intel and Nvidia GPUs. It would extend compatibility and it'd give an option for people who don't want the huge CUDA installation or want one installation they can use with GPUs of different vendors.
The problem is probably Torch not officially supporting Vulkan, though people are working on it, I would really hope for this to be a feature soon, even if experimental:
https://docs.pytorch.org/tutorials/prototype/vulkan_workflow.html
I don't know how much work it'd take to implement this.

**Comment by tuaris** [comment]
Vulkan support will also allow FreeBSD users to run GPU accelerated workloads without having to use the linuxulator since NVIDIA doesn't provide CUDA on FreeBSD.  Assuming of course we can even get ComfyUI to run on FreeBSD (most of the dependencies work).

**Comment by evcharger** [comment]
+1 for Vulkan, it will enable everyone with iGPUs from AMD to use comfyUI. Vulkan at the moment is 2x faster(tg and pp) than ROCm for LLM inferencing in the cases with AMD iGPUs under Linux.

**Comment by Juarrow** [comment]
+1

**Comment by nvdg** [comment]
Another +1 since Rocm is so shitty. On LLM's Vulkan is already faster than Rocm!

**Comment by GreenShadows** [comment]
Everyone agrees that implementing Vulkan is a good thing; for two years now, everyone has been nodding their heads in agreement.  :')

**Comment by MerijnHendriks** [comment]
Would love to see it!

**Comment by Tamalero** [comment]
This been sitting for years. Is there really any development for a vulkan backend?

**Comment by waltercool** [comment]
> This been sitting for years. Is there really any development for a vulkan backend?

Probably not, but feel free to send a commit with a proposal

**Comment by maikelthedev** [comment]
Please add vulkan

**Comment by vlforit** [comment]
Vulkan means freedom

**Comment by waltercool** [comment]
> Vulkan measn freedom

Vulkan is nice, but not the right answer. It's a good fallback system tho, like Triton kernel.

Using generic APIsl calls it's always going to be less efficient than native API calls. Even videogames with Vulkan use native Nvidia or AMD calls for certain features.

**Comment by GreenShadows** [comment]
> > Vulkan measn freedom
> 
> Vulkan is nice, but not the right answer. It's a good fallback system tho, like Triton kernel.
> 
> Using generic APIsl calls it's always going to be less efficient than native API calls. Even videogames with Vulkan use native Nvidia or AMD calls for certain features.

That's not true. 

Setting aside attention mechanisms, SD.cpp using Vulkan is very similar and sometimes faster than ComfyUI with CUDA or ROCm backend

**Comment by waltercool** [comment]
ROCm is not longer slower than Vulkan in the latest TheRock... but for sure, Vulkan is far more compatible.

Also, some models are starting to be optimized specifically for CUDA or ROCm, which is the case of NVFP4 and ROCMFP4, in the same way MLX gets huge improvements on Apple devices. It's a nice improvement from generic quant models (GGUF)

**Comment by GreenShadows** [comment]
INT8 and FP8 have been implemented in GGML. There’s nothing that ROCm or CUDA supports that Vulkan can’t. It’s just code.

**Comment by waltercool** [comment]
> INT8 and FP8 have been implemented in GGML. There’s nothing that ROCm or CUDA supports that Vulkan can’t. It’s just code.

It's code, running on different stacks. Cuda knows every single GPU, same ROCM HIP about AMD/Radeon GPUs.

Vulkan knows nothing, it's a generic set of instructions most GPU comply.

ROCm and Cuda are complex because it's intended for scientific purposes, Vulkan it's simpler, so easier to optimize, that's why it worked better on AMD in the last 3-4 years, but have it's own limits. Vulkan it's not designed for this purpose entirely 

Also, NVFP4 and ROCMFP4 are optimized model quants for dedicated frameworks, for specific GPU architectures. You get a considerable improvement (~10-30%) when used because the data structure (block layout) and scale format are optimized in the way Nvidia and AMD designed their stuff.

**Comment by GreenShadows** [comment]
And yet Vulkan is just as fast or even faster.

I just checked in Llama.cpp, with FA enabled on an RDNA3 GPU, and Vulkan is still 10% faster than ROCm.

**Comment by Juarrow** [comment]
"It works on my machine" has always been the best argument - ever... Of all time

**Comment by softcookiepp** [comment]
> > INT8 and FP8 have been implemented in GGML. There’s nothing that ROCm or CUDA supports that Vulkan can’t. It’s just code.
> 
> It's code, running on different stacks. Cuda knows every single GPU, same ROCM HIP about AMD/Radeon GPUs.
> 
> Vulkan knows nothing, it's a generic set of instructions most GPU comply.
> 
> ROCm and Cuda are complex because it's intended for scientific purposes, Vulkan it's simpler, so easier to optimize, that's why it worked better on AMD in the last 3-4 years, but have it's own limits. Vulkan it's not designed for this purpose entirely
> 
> Also, NVFP4 and ROCMFP4 are optimized model quants for dedicated frameworks, for specific GPU architectures. You get a considerable improvement (~10-30%) when used because the data structure (block layout) and scale format are optimized in the way Nvidia and AMD designed their stuff.

And yet if you want to use any of these optimized frameworks on older hardware, you will find that it is mostly unsupported. In NVIDIA's case, they dropped support for CUDA from any GPU older than Turing generation. In AMD's situation, things are even more dire.

So those without hardware supported by optimized frameworks would greatly benefit, as their alternative is to either use the CPU backend, buy a new GPU at 10x markup or whatever it is now, or run it on someone else's computer via cloud.

> And yet Vulkan is just as fast or even faster.
> 
> I just checked in Llama.cpp, with FA enabled on an RDNA3 GPU, and Vulkan is still 10% faster than ROCm.

Indeed. What the user above said isn't entirely correct. Vulkan has a *lot* of extension capability. For instance, most modern GPUs with tensor cores are compatible with the `VK_KHR_cooperative_matrix` extension, which allows Vulkan to use their tensor cores. Other data types commonly used in ML are supported at the hardware level via extensions too (fp16, bf16, fp8, etc.)

Anyways, ComfyUI depends on pytorch for the majority of its local functionality. Pytorch technically has a Vulkan backend, but it is unmaintained, incomplete, and not very portable. So I have spent the past half-year or so porting [pytorch_dlprim](https://github.com/artyom-beilis/pytorch_dlprim), a pytorch backend written in OpenCL, to Vulkan. My fork can be found [here.](https://github.com/softcookiepp/pytorch_dlprim) All operators implemented thus far are passing correctness tests, and now I am focused on trying to optimize it. Any help would be appreciated.

There are also several other individuals attempting to write Vulkan backends too:
https://github.com/Peterc3-dev/torch-vulkan
https://github.com/ZainSharief/torchvulkan
https://github.com/amitbar05/pytorch

Some are more complete than others. I intend to merge their efforts into my own, or coordinate with them at some point.
